torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,408 @@
+# Owner(s): ["oncall: distributed"]
+import collections
+import logging
+import operator
+import tempfile
+from dataclasses import dataclass
+from functools import wraps
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+import torch
+import torch.distributed as dist
+import torch.fx as fx
+from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
+from torch.distributed._spmd.graph_utils import (
+    CommType,
+    dump_graphs_to_files,
+    get_output,
+)
+from torch.distributed._spmd.iter_graph_module import IterGraphModule
+from torch.fx.passes.shape_prop import TensorMetadata
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+logger: logging.Logger = logging.getLogger("graph_optimization")
+aten = torch.ops.aten
+fake_tensor_mode = FakeTensorMode()
+
+_optimized_func: Set[str] = set()
+_run_before_sets: DefaultDict[str, Set[str]] = collections.defaultdict(set)
+_dump_graph_folder: str = ""
+
+
+logger: logging.Logger = logging.getLogger("graph_optimization")
+
+
+def enable_graph_optimization_dump(folder: str = ""):
+    global _dump_graph_folder
+    if not folder:
+        folder = tempfile.mkdtemp()
+    _dump_graph_folder = folder
+
+
+def graph_optimization_pass(
+    run_after: Iterable[str] = tuple(),
```

Comment:
Do we expect users to pass in run after names or is this mainly internal configuration (except for expert users who want to customize graph optimization)

Reply:
This is only for developers who develop graph optimization passes to specify the order constraint.